api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -173,18 +193,28 @@ def __next__(self) -> _SOMAChunk:
 
         # note: the `blockwise` call is employed for its ability to reindex the axes of the sparse matrix,
         # but the blockwise iteration feature is not used (block_size is set to retrieve the chunk as a single block)
-        scipy_iter = (
-            self.X.read(coords=(obs_joinids_chunk, self.var_joinids))
-            .blockwise(axis=0, size=len(obs_joinids_chunk), eager=False)
-            .scipy(compress=True)
+        blockwise_iter = self.X.read(coords=(obs_joinids_chunk, self.var_joinids)).blockwise(
+            axis=0, size=len(obs_joinids_chunk), eager=False
         )
-        X_batch, _ = next(scipy_iter)
+
+        if not self.return_sparse_X:
+            batch_iter = _tables_to_np(blockwise_iter.tables(), shape=(obs_batch.shape[0], len(self.var_joinids)))
+        else:
+            batch_iter = blockwise_iter.scipy(compress=True)
+
+        res = next(batch_iter)
+        X_batch: ChunkX = res[0]
+        if isinstance(X_batch, np.ndarray):
```

Comment:
nit: I would make this conditional on the same value used to select the iterator (e.g., `self.return_sparse_X`), and assert it is an ndarray.

Just for clarity - no other reason (i.e, the above code works AFAIK)

Reply:
Yeah, I wasn't totally happy with the logic on this.

The type of the value should be invariant with the value of `return_sparse_X`. But that's not how it was in the original PR or the previous code. But I can go for consistency with other branches here.

Reply:
Its sorta weird that we make an iterator here but only ever use the first iteration